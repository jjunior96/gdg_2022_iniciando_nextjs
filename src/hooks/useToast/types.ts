export type Props = {
  type?: 'success' | 'warning' | 'error';
  title: string;
};

export type ToastMessageProps = {
  id: string;
} & Props;

export type ToastDataContext = {
  addToast(message: Props): void;
  removeToast(id: string): void;
};

export type ToastProviderProps = {
  children: React.ReactNode;
};
