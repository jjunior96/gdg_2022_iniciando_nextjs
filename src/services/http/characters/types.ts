export type CharactersProps = {
  id: number;
  name: string;
  status: string;
  species: string;
  url: string;
  gender: string;
  image: string;
};

export type ReturnCharactersProps = {
  results: CharactersProps[];
};
