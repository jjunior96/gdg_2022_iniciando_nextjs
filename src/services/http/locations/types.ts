export type LocationsProps = {
  id: number;
  name: string;
  dimension: string;
  residents: string[];
  url: string;
};

export type ReturnLocationsProps = {
  results: LocationsProps[];
};
