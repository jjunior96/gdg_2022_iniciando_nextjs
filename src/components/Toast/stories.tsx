import { Story, Meta } from '@storybook/react/types-6-0';

import Toast, { ToastProps } from '.';

export default {
  title: 'Toast',
  component: Toast,
  decorators: [
    (Story) => (
      <div style={{ width: '90%', margin: '0 auto', position: 'relative' }}>
        <Story />
      </div>
    )
  ],
  argTypes: {
    messages: {
      type: ''
    }
  }
} as Meta;

export const Default: Story<ToastProps> = (args) => <Toast {...args} />;

Default.args = {
  messages: [
    {
      id: '1',
      type: 'success',
      title: 'Toast'
    }
  ]
};

export const withWarning: Story<ToastProps> = (args) => <Toast {...args} />;

withWarning.args = {
  messages: [
    {
      id: '1',
      type: 'warning',
      title: 'Toast'
    }
  ]
};

export const withError: Story<ToastProps> = (args) => <Toast {...args} />;

withError.args = {
  messages: [
    {
      id: '1',
      type: 'error',
      title: 'Toast'
    }
  ]
};
