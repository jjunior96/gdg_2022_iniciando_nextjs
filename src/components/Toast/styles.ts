import { ToastMessageProps } from 'hooks/useToast/types';
import styled, { css, DefaultTheme, keyframes } from 'styled-components';

type ToastTypeProps = Pick<ToastMessageProps, 'type'>;

export const Container = styled.div`
  ${({ theme }) => css`
    position: absolute;
    top: 0;
    right: 0;

    width: 35rem;

    z-index: ${theme.layers.alwaysOnTop};

    margin-top: ${theme.spacings.xxsmall};
    margin-right: ${theme.spacings.xxsmall};
  `}
`;

const contentModifiers = {
  success: (theme: DefaultTheme) => css`
    background-color: ${theme.colors.green};
  `,

  warning: (theme: DefaultTheme) => css`
    background-color: ${theme.colors.yellow_300};
  `,

  error: (theme: DefaultTheme) => css`
    background-color: ${theme.colors.red};
  `
};

const ContentAnimation = keyframes`
  from {
    opacity: 0;
    transform: translateX(50px);
  }

  to {
    opacity: 1;
    transform: translateX(0);
  }
`;

export const Content = styled.div<ToastTypeProps>`
  ${({ theme, type }) => css`
    font-size: ${theme.font.sizes.xsmall};

    color: ${theme.colors.black};

    font-weight: ${theme.font.bold};

    padding: ${theme.spacings.xxsmall};
    margin-bottom: ${theme.spacings.xxsmall};

    border-radius: ${theme.border.radius};

    position: relative;

    display: flex;
    align-items: center;

    animation: ${ContentAnimation} 0.5s;

    box-shadow: ${theme.box.shadow};

    > svg {
      margin-right: ${theme.spacings.xxxsmall};
    }

    ${!!type && contentModifiers[type](theme)};
  `}
`;

export const ButtonClose = styled.button`
  ${({ theme }) => css`
    border: 0;

    background-color: transparent;

    position: absolute;

    top: 0;
    right: 0;

    margin-top: ${theme.spacings.xxxsmall};
    margin-right: ${theme.spacings.xxxsmall};

    svg {
      color: ${theme.colors.white};

      transition: color 0.4s;

      &:hover {
        color: ${theme.colors.black};
      }
    }
  `}
`;
