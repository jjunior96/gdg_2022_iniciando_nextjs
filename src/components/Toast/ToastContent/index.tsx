import { useEffect, useState } from 'react';
import {
  FiAlertCircle as WarningIcon,
  FiCheckCircle as SuccessIcon
} from 'react-icons/fi';
import { RiCloseFill as CloseButton } from 'react-icons/ri';

import { useToast } from 'hooks/useToast';
import { ToastMessageProps } from 'hooks/useToast/types';

import * as S from './styles';

const ToastContent = ({ id, type, title }: ToastMessageProps) => {
  const [seconds, setSeconds] = useState(3000);
  const [pause, setPause] = useState(false);

  const { removeToast } = useToast();

  const icons = {
    success: <SuccessIcon size={20} />,
    warning: <WarningIcon size={20} />,
    error: <WarningIcon size={20} />
  };

  useEffect(() => {
    const timer = setInterval(() => {
      if (!pause && seconds > 0) {
        setSeconds(seconds - 1000);
        removeToast(id);
      }
    }, seconds);

    return () => {
      clearTimeout(timer);
    };
  }, [id, pause, removeToast, seconds]);

  return (
    <S.Container
      key={id}
      type={type}
      layout
      initial={{ opacity: 0, y: -50, scale: 0.3 }}
      animate={{ opacity: 1, y: 0, scale: 1 }}
      exit={{ opacity: 0, y: 20, scale: 0.5 }}
      onMouseOver={() => setPause(true)}
      onMouseLeave={() => setPause(false)}
    >
      {icons[type!]}

      <p>{title}</p>

      <S.ButtonClose type="button" onClick={() => removeToast(id)}>
        <CloseButton size={17} />
      </S.ButtonClose>
    </S.Container>
  );
};

export default ToastContent;
