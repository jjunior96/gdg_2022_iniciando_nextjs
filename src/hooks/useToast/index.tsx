import { createContext, useCallback, useContext, useState } from 'react';

import { v4 } from 'uuid';

import Toast from 'components/Toast';

import {
  Props,
  ToastDataContext,
  ToastMessageProps,
  ToastProviderProps
} from './types';

const ToastContext = createContext<ToastDataContext>({} as ToastDataContext);

export const ToastProvider = ({ children }: ToastProviderProps) => {
  const [messages, setMessages] = useState<ToastMessageProps[]>([]);

  const addToast = useCallback((message: Props) => {
    const id = v4();

    const toast = {
      id,
      ...message
    };

    setMessages((oldMessages) => [...oldMessages, toast]);
  }, []);

  const removeToast = useCallback(
    (id: string) => {
      const newMessages = messages.filter((item) => item.id !== id);

      setMessages(newMessages);
    },
    [messages]
  );

  return (
    <ToastContext.Provider
      value={{
        addToast,
        removeToast
      }}
    >
      {children}
      <Toast messages={messages} />
    </ToastContext.Provider>
  );
};

export default ToastContext;

export const useToast = () => {
  const context = useContext(ToastContext);

  if (!context) {
    throw new Error('useToast must be used within a ToastProvider');
  }

  return context;
};
