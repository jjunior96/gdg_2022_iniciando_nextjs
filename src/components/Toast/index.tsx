import { ToastMessageProps } from 'hooks/useToast/types';

import * as S from './styles';
import ToastContent from './ToastContent';

export type ToastProps = {
  messages: ToastMessageProps[];
};

const Toast = ({ messages }: ToastProps) => {
  return (
    <S.Container>
      {messages.map((item) => (
        <ToastContent
          key={item.id}
          id={item.id}
          title={item.title}
          type={item.type}
        />
      ))}
    </S.Container>
  );
};

export default Toast;
