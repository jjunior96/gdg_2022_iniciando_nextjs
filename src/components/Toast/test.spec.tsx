import { screen } from '@testing-library/react';
import ToastContext from 'hooks/useToast';
import { ToastMessageProps } from 'hooks/useToast/types';

import { renderWithTheme } from 'utils/tests/__config__/helpers';

import theme from 'styles/theme';

import Toast from '.';

const valuesMock = {
  addToast: jest.fn(),
  removeToast: jest.fn()
};

describe('<Toast />', () => {
  it('should render the Toast component', () => {
    const messages: ToastMessageProps[] = [
      {
        id: '1',
        type: 'success',
        title: 'Toast'
      }
    ];

    const { container } = renderWithTheme(
      <ToastContext.Provider value={valuesMock}>
        <Toast messages={messages} />;
      </ToastContext.Provider>
    );

    expect(screen.getByText(/toast/i)).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the when warning passed', () => {
    const messages: ToastMessageProps[] = [
      {
        id: '1',
        type: 'warning',
        title: 'Toast'
      }
    ];

    renderWithTheme(<Toast messages={messages} />);

    expect(screen.getByText(/toast/i).parentElement).toHaveStyle({
      'background-color': theme.colors.yellow_300
    });
  });

  it('should render the when error passed', () => {
    const messages: ToastMessageProps[] = [
      {
        id: '1',
        type: 'error',
        title: 'Toast'
      }
    ];

    renderWithTheme(<Toast messages={messages} />);

    expect(screen.getByText(/toast/i).parentElement).toHaveStyle({
      'background-color': theme.colors.red
    });
  });

  // it('should remove when user click on close button', async () => {
  //   const removeToast = jest.fn();

  //   const messages: ToastMessageProps[] = [
  //     {
  //       id: '1',
  //       type: 'error',
  //       title: 'Toast'
  //     }
  //   ];

  //   renderWithTheme(<Toast messages={messages} />);

  //   // NÃ£o deve chamar a funcao removeToast
  //   expect(removeToast).not.toHaveBeenCalled();

  //   const closeButton = screen.getByRole('button');

  //   fireEvent.click(closeButton);

  //   await waitFor(() => {
  //     expect(removeToast).toHaveBeenCalledTimes(1);
  //   });
  // });
});
