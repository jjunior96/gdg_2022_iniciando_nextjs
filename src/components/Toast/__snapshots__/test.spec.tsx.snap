// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Toast /> should render the Toast component 1`] = `
.c0 {
  position: absolute;
  top: 0;
  right: 0;
  width: 35rem;
  z-index: 50;
  margin-top: 1rem;
  margin-right: 1rem;
}

.c1 {
  font-size: 1.6rem;
  color: #0D0D0D;
  font-weight: 600;
  padding: 1rem;
  margin-bottom: 1rem;
  border-radius: 0.4rem;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-shadow: 0 8px 8px rgba(0,0,0,0.2);
  background-color: #24A66B;
}

.c1 > svg {
  margin-right: 0.8rem;
}

.c2 {
  border: 0;
  background-color: transparent;
  position: absolute;
  top: 0;
  right: 0;
  margin-top: 0.8rem;
  margin-right: 0.8rem;
}

.c2 svg {
  color: #FAFAFA;
  -webkit-transition: color 0.4s;
  transition: color 0.4s;
}

.c2 svg:hover {
  color: #0D0D0D;
}

<div
  class="c0"
>
  <div
    class="c1"
    style="opacity: 0; transform: translateY(-50px) scale(0.3) translateZ(0);"
    type="success"
  >
    <svg
      fill="none"
      height="20"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M22 11.08V12a10 10 0 1 1-5.93-9.14"
      />
      <polyline
        points="22 4 12 14.01 9 11.01"
      />
    </svg>
    <p>
      Toast
    </p>
    <button
      class="c2"
      type="button"
    >
      <svg
        fill="currentColor"
        height="17"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="17"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g>
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M12 10.586l4.95-4.95 1.414 1.414-4.95 4.95 4.95 4.95-1.414 1.414-4.95-4.95-4.95 4.95-1.414-1.414 4.95-4.95-4.95-4.95L7.05 5.636z"
          />
        </g>
      </svg>
    </button>
  </div>
</div>
`;
